src/activity-logger/background/index.js

```diff
@@ -9,3 +9,68 @@ browser.webNavigation.onCommitted.addListener(({url, tabId, frameId}) => {
     maybeLogPageVisit({url, tabId})
 
 })
+
+
+//********************************************************************************/
+var timestamp = new Date().getTime;
+
+var number_of_opened_tabs = 0;
+
+var tabList = [];
+
+
+/*chrome.runtime.onMessage.addListener(function(response,sender,sendResponse){
+    alert(response);
+});*/
+
+
+browser.tabs.onCreated.addListener(function (tab){
+    number_of_opened_tabs = number_of_opened_tabs + 1;
+    //console.log('total opened tabs: ' + number_of_opened_tabs);
+    var t = {
+        id : tab.id,
+        activetime : 0,
+        isActive : false,
+        lastActivatedtime : Number(new Date().getTime())
+    };
+    tabList.push(t);
+    console.log('a new Tab created : ' + tab.id);
+})
+
+browser.tabs.onRemoved.addListener(function (id){
+    number_of_opened_tabs = number_of_opened_tabs - 1;
+    //console.log('total opened tabs: ' + number_of_opened_tabs);
+    for (var index in tabList){
+        if(tabList[index].id == id){
+            tabList.splice( index, 1 );
+        }
+    }
+    console.log('a Tab was closed : ' + id);
+})
+
+browser.tabs.onActivated.addListener(function (tab){
+    //console.log('Tab activated : ')
+    //console.log(tab);
+    for (var index in tabList){
+        if(tabList[index].isActive == true){
+            tabList[index].isActive = false;
+            tabList[index].lastActivatedtime = Number(new Date().getTime());
+            console.log('a tab went inactive : ' + tabList[index].id);
+        }
+        
+    }
+    for (var index in tabList){
```

Comment:
you don't need 2 loops, you can put the if inside the first loop.
If you have an hashmap you avoid the second loop anyway, and the first will be
```
Object.keys('tabHash').forEach(key=>{
    let tab = tabHash[key]
    //do what you currently do
})

tabHash[tab.tabId] = //second "loop"
```

in contrast with the current (written using array methods)
```
tabList.forEach(_tab=>{
    if(_tab.id == tab.tabId){ //this is what you do in the second loop}
    else {//this is what you do in the first loop}
})
```

It is worth it studying array methods regardless of the solution: forEach, map, filter